vllm/utils.py

```diff
@@ -1323,16 +1326,78 @@ def add_arguments(self, actions):
         super().add_arguments(actions)
 
 
+class _FlexibleArgumentGroup(_ArgumentGroup):
+
+    def __init__(self, parser: FlexibleArgumentParser, *args, **kwargs):
+        self._parser = parser
+        super().__init__(*args, **kwargs)
+
+    def add_argument(self, *args: Any, **kwargs: Any):
+        if sys.version_info < (3, 13):
+            deprecated = kwargs.pop('deprecated', False)
+            action = super().add_argument(*args, **kwargs)
+            object.__setattr__(action, 'deprecated', deprecated)
+            if deprecated and action.dest not in \
+                    self._parser.__class__._deprecated:
+                self._parser._warning(
+                    _gettext("argument '%(argument_name)s' is deprecated") %
+                    {'argument_name': action.dest})
+                self._parser._deprecated.add(action.dest)
+            return action
+
+        # python>3.13
+        return super().add_argument(*args, **kwargs)
+
+
 class FlexibleArgumentParser(ArgumentParser):
     """ArgumentParser that allows both underscore and dash in names."""
 
+    _deprecated: set[str] = set()
+
     def __init__(self, *args, **kwargs):
         # Set the default 'formatter_class' to SortedHelpFormatter
         if 'formatter_class' not in kwargs:
             kwargs['formatter_class'] = SortedHelpFormatter
         super().__init__(*args, **kwargs)
 
-    def parse_args(self, args=None, namespace=None):
+    def add_argument(self, *args: Any, **kwargs: Any):
+        # add a deprecated=True with optional deprecated_reason to signify
+        # reasons for deprecating this args
+        if sys.version_info < (3, 13):
+            deprecated = kwargs.pop('deprecated', False)
+            action = super().add_argument(*args, **kwargs)
+            object.__setattr__(action, 'deprecated', deprecated)
+            if deprecated and \
+                action.dest not in FlexibleArgumentParser._deprecated:
+                self._warning(
+                    _gettext("argument '%(argument_name)s' is deprecated") %
+                    {'argument_name': action.dest})
+                self._deprecated.add(action.dest)
```

Comment:
What's the purpose of the class variable `_deprecated`? Is it a way of ensuring that each deprecated arg is only logged once across all instances of `FlexibleArgumentParser`?

Reply:
yes. We imports FlexibleParserArgument multiple times in many places